### PR TITLE
control sequences inclusion

### PR DESCRIPTION
**Pull request by eabderh**
changed tty.setraw to tty.setcbreak so that control sequences (CTRL-C)
are processed by the console. fixed unescapable loop bug.

example of bug:
>>> from getch import pause
>>> while True: pause()

this loop can't be escaped with CTRL-C if tty.setraw is turned on. with
setcbreak it will exit on CTRL-C.

lets have a good start to 2017!

### CODE REVIEW

**Comment by joeyespo** [comment]
This looks good to me.

One question though. Doing some research led me to [this article](https://utcc.utoronto.ca/~cks/space/blog/unix/CBreakAndRaw). Looks like there's [a lot of different flags](https://hg.python.org/cpython/file/618ea5612e83/Lib/tty.py) between the two modes. There's also a dedicated flag called `ISIG` which "with this disabled, things like ^C do not generate interrupts when they're typed." Would it be better to call `setraw` then adjust that one flag instead?

I don't know enough to know. I'm sure there's a good reason `setcbreak` doesn't just set that one flag and instead does all that other stuff. If you recommend `setcbreak`, I'll pull this in. ([Here's the question that led me to that article.](http://stackoverflow.com/questions/12231794/python-in-raw-mode-stdin-print-adds-spaces) It also recommends just using `setcbreak`.)

Thanks!

**Comment by eabderh** [comment]
Thanks for doing all that research!

After doing my own research I'm of the opinion that `setraw` should be used __as default__ in the `getch` function, but that `getch` should have an optional argument on the linux side of things that lets `setcbreak` be set.

This doesn't break compatibility and keeps the main `getch` functionality the same as the msvcrt `getch` function. A [link](https://docs.python.org/2/library/msvcrt.html#msvcrt.getch) to description:
> ... The Control-C keypress cannot be read with this function...

The `pause` function can use `getch` with the `setcbreak` flag set, so it, by default, will have the expected behavior of following control sequences and the bug won't be an issue.

As for the optional flags, I recommend keeping it between `setraw` and `setcbreak`. They both behave as expected and I couldn't find any use in defining other variations.

In conclusion, if you are in agreement with the above I can get started on an updated version of `getch` with the `setcbreak` flag. This pull request should probably be not included though and can be closed.

**Comment by joeyespo** [comment]
Actually, I nearly forgot. In #4, it turns out that `getch()` will return `'\x03'` on Ctrl+C. Could you verify that this works for you too?

With that, I bet we can make a cross-platform solution for `pause`:

```py
def pause(message='Press any key to continue . . . '):
    """
    Prints the specified message if it's not None and waits for a keypress.
    """
    if message is not None:
        print(message, end='')
        sys.stdout.flush()
    ch = getch()
    print()
    if ch == '\x03':
        raise KeyboardInterrupt()
```

Then you can break out of your original code snippet and keep `getch` working as-is.

(In my opinion, fewer arguments / customizations is a good thing if you can solve everything well enough without them. Even if they have sensible defaults, each optional argument is another decision a user ultimately needs to make.)

What do you think?

**Comment by eabderh** [comment]
Well, not checking the previous issues was very stupid of me. Yes, #4 does solve my original issue.

Just being pedantic here, but there is a small issue with the solution. By disabling the __ISIG__ flag from the [termios man pages](http://man7.org/linux/man-pages/man3/termios.3.html) in setraw ([link](https://hg.python.org/cpython/file/618ea5612e83/Lib/tty.py)) the ability to suspend a process is also disabled, along with other control sequences. So, the CTRL-Z (`'\x1a'`) sequence will not work and it doesn't seem to be very straight forward to implement this functionality in the same way as in #4.

Again, this isn't really an issue and fixing it would cause all kinds of problems with compatibility, especially with `msvcrt`. The only thing I would recommend is to change the `pause` function to check for CTRL-C as in #4 or maybe include it in the README. If you would like, I can make the changes myself and resubmit them.

**Comment by joeyespo** [comment]
> the ability to suspend a process is also disabled, along with other control sequences.

Ah, that's true.

I'm not sure what to do about these points yet. If Windows doesn't have these same issues, it might be worth switching to `setcbreak`. To know, we'll have to do some more digging to see why `setraw` was chosen over `setcbreak` in the first place. Or rephrased, is the `setcbreak` implementation still equivalent in a cross-platform way? Or are there breaking differences? I don't know enough yet.

As for the Ctrl+C part, go for it. I'll pull in a PR that fixes `pause` to raise `KeyboardInterrupt` so at least you can break out of infinite loops.

**Comment by eabderh** [comment]
Ok, changes made with (hopefully) minimal upheaval. Most of the changes are documented in the commit messages but the general updates are the addition of CTRL-C handling in `pause` and some improvement in checking for `ns` or `posix` systems.

> ...Or rephrased, is the setcbreak implementation still equivalent in a cross-platform way?

In my opinion, no. I've tried to keep the functionality of both the linux and the windows version of `getch` compatible with each other.

The only issue, which I noted in commit 40b2bff, is that the windows `getch` from `msvcrt` returns a `bytes` string literal (something like `b'\x03'` instead of `'\x03'`). This can be fixed by adding a shim to the windows `getch` function that converts the output to a `str` through `chr(ord(getch()))` or `getch().decode()`

_Testing done with Python 3.6.0 on Windows 7 (64bit)_

**Comment by muammar** [comment]
Just to add something about this. When catching `\x03` I have realized that the terminal gets broken. For instance, if I am running an app with py-getch, and I kill it with what's suggested in #4, then it turns out that the terminal does not work properly e.g. running sudo and trying to enter the password does nothing. Does this PR causes this as well?.

**Comment by eabderh** [comment]
In response to @muammar :

> ...the terminal does not work properly e.g. running sudo and trying to enter the password does nothing.

I am assuming you are running linux, is that right? Just in case this an issue of a misunderstanding, linux does __not__ show passwords when they are entered. Here is the [stack overflow question](http://unix.stackexchange.com/questions/268006/why-is-the-password-i-entered-not-visible/268013) about this.

Let me know if this is the cause, otherwise I can help you more if you can give me more information (what else on the terminal breaks, what system you are using, versions, etc). I have tried multiple times to get the same problem as you but my terminal did not break, so regrettably I can't help you right away :(

You might also want to open a new issue as a PR isn't the best place to discuss this issue. Like I said, my PR doesn't seem to be correlated with your problem. Honestly though, I'm new to github and PRs so let me know if this is the right/wrong way of doing things. Thanks!

**Comment by muammar** [comment]
> I am assuming you are running linux, is that right? Just in case this an issue of a misunderstanding, linux does not show passwords when they are entered. Here is the stack overflow question about this.

I am using linux since more than 12 years ago :). What I mean is that if you enter your password and press enter nothing happens. It seems like if enter is not being registered by the terminal after you kill an application using py-getch with `Ctrl-C`. I hope I have made myself clear. Besides that, if for instance, you execute `ls` the format of the rendered text is messed up. The only way to recover your terminal is to execute `reset` (or open a new session, of course).

**Comment by eabderh** [comment]
> I am using linux since more than 12 years ago :

Sorry, did not mean to offend :) Just wanted to cover all bases.

Ok, here is my thinking. Since `reset` fixed the issue, `reset` must set the [flags in termios](http://man7.org/linux/man-pages/man3/termios.3.html) that were not reset in the `getch` function. One way to check this could be by using the `stty -a` command in the `bash` shell, which prints out all these flags in human readable form. Run the command before and after running the program with the issue and see if they both give out the same output, which they should. Hope this helps.

Just as a note, the "press enter nothing happens" could be the `ICRNL` termios flag, which has to do with converting carriage returns into newlines and is unset in `getch`.

**Comment by muammar** [comment]
>Sorry, did not mean to offend :) Just wanted to cover all bases.

No offence taken at all :). I wrote in a rush, and not being native english speaker, I wasn't able to express myself correctly. I will try to reproduce this problem with an [app](https://github.com/muammar/mkchromecast) I made that uses this module and has this problem.

>Ok, here is my thinking. Since reset fixed the issue, reset must set the flags in termios that were not reset in the getch function. One way to check this could be by using the stty -a command in the bash shell, which prints out all these flags in human readable form. Run the command before and after running the program with the issue and see if they both give out the same output, which they should. Hope this helps.

Thanks for this information. I will try that although I am using `zsh`. 

> Just as a note, the "press enter nothing happens" could be the ICRNL termios flag, which has to do with converting carriage returns into newlines and is unset in getch.

Oh, that makes lots of sense. Maybe @joeyespo can ask that.
